dcos/http.py

```diff
@@ -146,37 +146,33 @@ def _request_with_auth(response,
     :rtype: requests.Response
     """
 
-    i = 0
-    while i < 3 and response.status_code == 401:
-        parsed_url = urlparse(url)
-        hostname = parsed_url.hostname
-        auth_scheme, realm = get_auth_scheme(response)
-        creds = (hostname, auth_scheme, realm)
-
-        with lock:
-            if creds not in AUTH_CREDS:
-                auth = _get_http_auth(response, parsed_url, auth_scheme)
-            else:
-                auth = AUTH_CREDS[creds]
-
-        # try request again, with auth
-        response = _request(method, url, is_success, timeout, auth,
-                            verify, **kwargs)
-
-        # only store credentials if they're valid
-        with lock:
-            if creds not in AUTH_CREDS and response.status_code == 200:
-                AUTH_CREDS[creds] = auth
-            # acs invalid token
-            elif response.status_code == 401 and \
-                    auth_scheme in ["acsjwt", "oauthjwt"]:
-
-                if config.get_config_val("core.dcos_acs_token") is not None:
-                    msg = ("Your core.dcos_acs_token is invalid. "
-                           "Please run: `dcos auth login`")
-                    raise DCOSException(msg)
+    parsed_url = urlparse(url)
+    hostname = parsed_url.hostname
+    auth_scheme, realm = get_auth_scheme(response)
+    creds = (hostname, auth_scheme, realm)
 
-        i += 1
+    with lock:
+        if creds not in AUTH_CREDS:
+            auth = _get_http_auth(response, parsed_url, auth_scheme)
+        else:
+            auth = AUTH_CREDS[creds]
+
+    # try request again, with auth
+    response = _request(method, url, is_success, timeout, auth,
+                        verify, **kwargs)
+
+    # only store credentials if they're valid
+    with lock:
+        if creds not in AUTH_CREDS and response.status_code == 200:
+            AUTH_CREDS[creds] = auth
+        # acs invalid token
+        elif response.status_code == 401 and \
+                auth_scheme in ["acsjwt", "oauthjwt"]:
+
+            if config.get_config_val("core.dcos_acs_token") is not None:
+                msg = ("Your core.dcos_acs_token is invalid. "
+                       "Please run: `dcos auth login`")
+                raise DCOSException(msg)
 
     if response.status_code == 401:
         raise DCOSAuthenticationException(response)
```

Comment:
I left it there because the prev implementation was responsible for throwing it ie I've preserved its contract. I don't think it is good to rely on the outer behaviour unless otherwise specified in its contract. 

That said, I'll change it if that's what you want. Please confirm.